File: src/questforge/core/contracts/reasoning_contract_v1.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional


CONTRACT_VERSION = "reasoning_contract_v1"


ReasoningLevel = Literal["weak", "ok", "good"]
EndTag = Literal["ending_result", "ending_wrong", "epilogue", "ending_check"]


@dataclass(frozen=True)
class ReasoningSummaryV1:
    version: str = CONTRACT_VERSION

    # context
    case_title: str = ""
    node_id: str = ""
    tag: EndTag = "ending_check"
    turn: int = 0

    # accuse / reason
    accused: str = ""  # suspect id or name
    reason_mode: str = "choice"  # choice/text/voice
    reason_ids: List[str] = field(default_factory=list)
    selected_observations: List[str] = field(default_factory=list)
    reason_text: str = ""
    reason_summary: str = ""

    # evidence
    clues_preview: List[str] = field(default_factory=list)

    # scoring
    level: ReasoningLevel = "weak"
    score: int = 0
    threshold: int = 0
    engine_message: str = ""

    matched_evidence: List[str] = field(default_factory=list)
    missing_key_evidence: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "version": self.version,
            "case_title": self.case_title,
            "node_id": self.node_id,
            "tag": self.tag,
            "turn": int(self.turn),
            "accused": self.accused,
            "reason_mode": self.reason_mode,
            "reason_ids": list(self.reason_ids),
            "selected_observations": list(self.selected_observations),
            "reason_text": self.reason_text,
            "reason_summary": self.reason_summary,
            "clues_preview": list(self.clues_preview),
            "level": self.level,
            "score": int(self.score),
            "threshold": int(self.threshold),
            "engine_message": self.engine_message,
            "matched_evidence": list(self.matched_evidence),
            "missing_key_evidence": list(self.missing_key_evidence),
        }


def _str_list(v: Any) -> List[str]:
    if not isinstance(v, list):
        return []
    out: List[str] = []
    for x in v:
        s = str(x).strip()
        if s:
            out.append(s)
    return out

def ensure_reasoning_contract_v1(meta: Dict[str, Any], *, defaults: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """把現有 meta normalize 成 contract v1 的 key set（缺的補預設）。"""
    defaults = defaults or {}
    out: Dict[str, Any] = {}
    out["version"] = CONTRACT_VERSION

    def pick(key: str, fallback: Any) -> Any:
        if key in meta and meta[key] is not None:
            return meta[key]
        if key in defaults and defaults[key] is not None:
            return defaults[key]
        return fallback

    out["case_title"] = str(pick("case_title", ""))
    out["node_id"] = str(pick("node_id", ""))
    out["tag"] = str(pick("tag", "ending_check"))
    out["turn"] = int(pick("turn", 0) or 0)

    out["accused"] = str(pick("accused", ""))
    out["reason_mode"] = str(pick("reason_mode", "choice"))
    out["reason_ids"] = _str_list(pick("reason_ids", []))
    out["selected_observations"] = _str_list(pick("selected_observations", []))
    out["reason_text"] = str(pick("reason_text", ""))
    out["reason_summary"] = str(pick("reason_summary", ""))

    out["clues_preview"] = _str_list(pick("clues_preview", []))

    out["level"] = str(pick("level", "weak"))
    out["score"] = int(pick("score", 0) or 0)
    out["threshold"] = int(pick("threshold", 0) or 0)
    out["engine_message"] = str(pick("engine_message", ""))

    out["matched_evidence"] = _str_list(pick("matched_evidence", []))
    out["missing_key_evidence"] = _str_list(pick("missing_key_evidence", []))

    return out

File: docs/flutter_contract/reasoning_contract_v1.dart
// lib/core/contracts/reasoning_contract_v1.dart
import 'dart:convert';

/// contract version (engine)
const kReasoningContractV1 = 'reasoning_contract_v1';

enum ReasoningLevel { weak, ok, good, unknown }
enum EndTag { endingResult, endingWrong, epilogue, endingCheck, unknown }

ReasoningLevel reasoningLevelFrom(String? s) {
  switch ((s ?? '').trim()) {
    case 'weak':
      return ReasoningLevel.weak;
    case 'ok':
      return ReasoningLevel.ok;
    case 'good':
      return ReasoningLevel.good;
    default:
      return ReasoningLevel.unknown;
  }
}

EndTag endTagFrom(String? s) {
  switch ((s ?? '').trim()) {
    case 'ending_result':
      return EndTag.endingResult;
    case 'ending_wrong':
      return EndTag.endingWrong;
    case 'epilogue':
      return EndTag.epilogue;
    case 'ending_check':
      return EndTag.endingCheck;
    default:
      return EndTag.unknown;
  }
}

int _asInt(dynamic v, {int fallback = 0}) {
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v.trim()) ?? fallback;
  return fallback;
}

String _asString(dynamic v, {String fallback = ''}) {
  if (v == null) return fallback;
  return v.toString();
}

List<String> _asStringList(dynamic v) {
  if (v is List) {
    return v.where((e) => e != null).map((e) => e.toString()).toList();
  }
  return const [];
}

class ReasoningSummaryV1 {
  final String version;

  // context
  final String caseTitle;
  final String nodeId;
  final EndTag tag;
  final int turn;

  // accuse / reason
  final String accused; // suspect id or name
  final String reasonMode; // choice/text/voice
  final List<String> reasonIds;
  final List<String> selectedObservations;
  final String reasonText;
  final String reasonSummary;

  // evidence
  final List<String> cluesPreview;

  // scoring
  final ReasoningLevel level;
  final int score;
  final int threshold;
  final String engineMessage;

  final List<String> matchedEvidence;
  final List<String> missingKeyEvidence;

  const ReasoningSummaryV1({
    required this.version,
    required this.caseTitle,
    required this.nodeId,
    required this.tag,
    required this.turn,
    required this.accused,
    required this.reasonMode,
    required this.reasonIds,
    required this.selectedObservations,
    required this.reasonText,
    required this.reasonSummary,
    required this.cluesPreview,
    required this.level,
    required this.score,
    required this.threshold,
    required this.engineMessage,
    required this.matchedEvidence,
    required this.missingKeyEvidence,
  });

  factory ReasoningSummaryV1.fromMap(Map<String, dynamic> map) {
    return ReasoningSummaryV1(
      version: _asString(map['version']),
      caseTitle: _asString(map['case_title']),
      nodeId: _asString(map['node_id']),
      tag: endTagFrom(_asString(map['tag'])),
      turn: _asInt(map['turn']),
      accused: _asString(map['accused']),
      reasonMode: _asString(map['reason_mode'], fallback: 'choice'),
      reasonIds: _asStringList(map['reason_ids']),
      selectedObservations: _asStringList(map['selected_observations']),
      reasonText: _asString(map['reason_text']),
      reasonSummary: _asString(map['reason_summary']),
      cluesPreview: _asStringList(map['clues_preview']),
      level: reasoningLevelFrom(_asString(map['level'])),
      score: _asInt(map['score']),
      threshold: _asInt(map['threshold']),
      engineMessage: _asString(map['engine_message']),
      matchedEvidence: _asStringList(map['matched_evidence']),
      missingKeyEvidence: _asStringList(map['missing_key_evidence']),
    );
  }

  /// 支援後端回傳 meta 可能是 Stringified JSON 的狀況（保守）
  static ReasoningSummaryV1? tryParse(dynamic meta) {
    if (meta == null) return null;

    if (meta is Map<String, dynamic>) {
      return ReasoningSummaryV1.fromMap(meta);
    }
    if (meta is Map) {
      return ReasoningSummaryV1.fromMap(meta.map((k, v) => MapEntry(k.toString(), v)));
    }
    if (meta is String) {
      final s = meta.trim();
      if (s.isEmpty) return null;
      try {
        final obj = jsonDecode(s);
        if (obj is Map) {
          return ReasoningSummaryV1.fromMap(obj.map((k, v) => MapEntry(k.toString(), v)));
        }
      } catch (_) {}
    }
    return null;
  }

  bool get isV1 => version.trim() == kReasoningContractV1;

  @override
  String toString() => 'ReasoningSummaryV1(tag=$tag level=$level score=$score/$threshold)';
}
